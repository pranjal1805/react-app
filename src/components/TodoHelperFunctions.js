import React, { useState } from 'react';

export const addTodo = (list, item) => [item, ...list]

export const addTodo1 = todo => {
    if (!todo.text || /^\s*$/.test(todo.text)) {
      return;
    }

    //Here we add our tasks to an array of already exisisting tasks
    const newTodos = [todo, ...todos];
  
    setTodos(newTodos);
    console.log(...todos);
};

export const randomId = () => Math.floor(Math.random() * 100000)

export const toggleTodo = (todo) => ({ ...todo, isComplete: !todo.isComplete })

export const updateTodo = (list, updatedItem) => { 
    const updatedIndex = list.findIndex(item => item.id === updatedItem.id)

    return [
        ...list.slice(0, updatedIndex),
        updatedItem,
        ...list.slice(updatedIndex + 1)
    ]
}

export const removeTodo = (list, todoId) => {
    const todoIndex = list.findIndex(item => item.id === todoId)
    return [
        ...list.slice(0, todoIndex),
        ...list.slice(todoIndex + 1)
    ]
}

export const search = (id, list) => list.find(item => item.id === id)
