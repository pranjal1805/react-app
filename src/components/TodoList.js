import React, { useState } from 'react';
import TodoForm from './TodoForm';
import Todo from './Todo';

function TodoList() {
  const [todos, setTodos] = useState([]);

  const addTodo = todo => {
    if (!todo.text || /^\s*$/.test(todo.text)) {
      return;
    }

    //Here we add our tasks to an array of already exisisting tasks
    const newTodos = [todo, ...todos];
  
    setTodos(newTodos);
    console.log(...todos);
  };

  const updateTodo = (todoId, newValue) => {
    if (!newValue.text || /^\s*$/.test(newValue.text)) {
      return;
    }

    //We make a new map of values. If the item id is equal to our todoID,
    //we update the map with a new value else we re-enter the same item in the map.
    setTodos(prev => prev.map(item => (item.id === todoId ? newValue : item)));
  };

  const removeTodo = id => {
    //the removedArr would have all tasks that don't have the same id as the Id passed as a parameter
    //This new array is passed to the use state.
    const removedArr = [...todos].filter(todo => todo.id !== id);

    setTodos(removedArr);
  };

  const completeTodo = id => {
    //Once the task is complete we strike it out
    //Thus to keep a check we toggle the isComplete
    let updatedTodos = todos.map(todo => {
      if (todo.id === id) {
        todo.isComplete = !todo.isComplete;
      }
      return todo;
    });
    setTodos(updatedTodos);
  };

  return (
    <>
      <h1>ToDo List!</h1>
      <TodoForm onSubmit={addTodo} />
      <Todo
        todos={todos}
        completeTodo={completeTodo}
        removeTodo={removeTodo}
        updateTodo={updateTodo}
      />
    </>
  );
}

export default TodoList;