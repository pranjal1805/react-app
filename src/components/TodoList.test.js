import React from 'react'
import renderer from 'react-test-renderer'
import TodoList from './TodoList'
import { addTodo, addTodo1, randomId, toggleTodo, updateTodo, removeTodo, search } from './TodoHelperFunctions'
import { addTodoFunction } from './TodoList'

test('addTodo should add a task to the list', ()=>{
    const startTodos = [
        { id: 1, name: 'one', isComplete: false },
        { id: 2, name: 'two', isComplete: false }
    ]

    const newTodo = { id: 3, name: 'three', isComplete: false }

    const expected = [
        { id: 3, name: 'three', isComplete: false },
        { id: 1, name: 'one', isComplete: false },
        { id: 2, name: 'two', isComplete: false }
    ]

    const result = addTodo(startTodos, newTodo)
    
    expect(result).toEqual(expected)
})

test('addTodo should not change the original order of the array', () => { 
    const startTodos = [
        { id: 1, name: 'one', isComplete: false },
        { id: 2, name: 'two', isComplete: false }
    ]

    const newTodo = { id: 3, name: 'three', isComplete: false }
    
    const expected = [
        { id: 1, name: 'one', isComplete: false },
        { id: 2, name: 'two', isComplete: false },
        { id: 3, name: 'three', isComplete: false }
    ]
 
    const result = addTodo(startTodos, newTodo)

    expect(result).not.toBe(startTodos)
})

test('Search should return an item present in the array', () => { 
    const startTodos = [
        { id: 1, name: 'one', isComplete: false },
        { id: 2, name: 'two', isComplete: false },
        { id: 3, name: 'three', isComplete: false }
    ]

    const expected = { id: 2, name: 'two', isComplete: false }

    const result = search(2, startTodos)

    expect(result).toEqual(expected)
})

test('this function toggles between complete and not complete', () => {
    const startTodo = { id: 1, name: 'one', isComplete: false }
    const expected = { id: 1, name: 'one', isComplete: true }

    const result = toggleTodo(startTodo)

    expect(result).toEqual(expected)
})

test('toggleTodo should not change the list of tasks', () => { 
    const startTodo = { id: 1, name: 'one', isComplete: false }
    const result = toggleTodo(startTodo)

    expect(result).not.toBe(startTodo)
})

test('updateTodo updates the tasks', () => { 
    const startTodos = [
        { id: 1, name: 'one', isComplete: false },
        { id: 2, name: 'two', isComplete: false },
        { id: 3, name: 'three', isComplete: false }
    ]

    const updatedTodo = { id: 2, name: 'two', isComplete: true }

    const expectedTodo = [
        { id: 1, name: 'one', isComplete: false },
        { id: 2, name: 'two', isComplete: true },
        { id: 3, name: 'three', isComplete: false }
    ]

    const result = updateTodo(startTodos, updatedTodo)

    expect(result).toEqual(expectedTodo)
})

test('updateTodo shouldn\'t change the ordering of the array', () => { 
    const startTodos = [
        { id: 1, name: 'one', isComplete: false },
        { id: 2, name: 'two', isComplete: false },
        { id: 3, name: 'three', isComplete: false }
    ]

    const updatedTodo = { id: 2, name: 'two', isComplete: true }

    const result = updateTodo(startTodos, updatedTodo)

    expect(result).not.toBe(startTodos)
})

test('Used for deletion', () => {
    const startTodos = [
        { id: 1, name: 'one', isComplete: false },
        { id: 2, name: 'two', isComplete: false },
        {id: 3, name: 'three', isComplete: false}
    ]

    const expected = [
        { id: 1, name: 'one', isComplete: false },
        {id: 3, name: 'three', isComplete: false}
    ]

    const result = removeTodo(startTodos, 2)

    expect(result).toEqual(expected)
})

test('removeTodo shouldn\'t change the ordering', () => { 
    const startTodos = [
        { id: 1, name: 'one', isComplete: false },
        { id: 2, name: 'two', isComplete: false },
        {id: 3, name: 'three', isComplete: false}
    ]

    const expected = [
        { id: 1, name: 'one', isComplete: false },
        {id: 3, name: 'three', isComplete: false}
    ]

    const result = removeTodo(startTodos, 2)

    expect(result).not.toBe(expected)
})