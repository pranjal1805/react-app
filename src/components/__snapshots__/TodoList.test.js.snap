// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders correctly 1`] = `
Array [
  <h1>
    ToDo List!
  </h1>,
  <form
    className="todo-form"
    onSubmit={[Function]}
  >
    <input
      className="todo-input"
      name="text"
      onChange={[Function]}
      placeholder="Add a todo"
      value=""
    />
    <button
      className="todo-button"
      onClick={[Function]}
    >
      Add Todo
    </button>
  </form>,
]
`;
